pkg/attestlib/jwt.go

```diff
@@ -45,6 +45,65 @@ func getAlgName(alg SignatureAlgorithm) string {
 	}
 }
 
+type jwtSigner struct {
+	privateKey         interface{}
+	publicKeyID        string
+	signatureAlgorithm SignatureAlgorithm
+}
+
+// NewJwtSigner creates a Signer interface for JWT Attestations. `publicKeyID`
+// is the ID of the public key that can verify the Attestation signature.
+// TODO: Explain formatting of JWT private keys.
+func NewJwtSigner(privateKey []byte, publicKeyID string, alg SignatureAlgorithm) (Signer, error) {
+	key, err := parsePkixPrivateKeyPem(privateKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing private key")
+	}
+
+	// If no ID is provided one is computed based on the default digest-based URI extracted from the public key material
+	if len(publicKeyID) == 0 {
+		publicKeyID, err = generatePkixPublicKeyId(key)
+		if err != nil {
+			return nil, errors.Wrap(err, "error generating public key id")
+		}
+	}
+	return &jwtSigner{
+		privateKey:         key,
+		publicKeyID:        publicKeyID,
+		signatureAlgorithm: alg,
+	}, nil
+}
+
+// CreateAttestation creates a signed JWT Attestation. See Signer for more details.
+func (s *jwtSigner) CreateAttestation(payload []byte) (*Attestation, error) {
+	type headerTemplate struct {
+		typ, alg, kid string
+	}
+	header := headerTemplate{
+		typ: "JWT",
+		alg: getAlgName(s.signatureAlgorithm),
+		kid: s.publicKeyID,
+	}
+
+	headerJson, err := json.Marshal(header)
+	if err != nil {
+		return nil, errors.Wrap(err, "error marshaling header")
+	}
+	headerBase64 := base64.RawURLEncoding.EncodeToString(headerJson)
+	payloadBase64 := base64.RawURLEncoding.EncodeToString(payload)
+	signature, err := createDetachedSignature(s.privateKey, []byte(headerBase64+"."+payloadBase64), s.signatureAlgorithm)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating signature")
+	}
+	signatureBase64 := base64.RawURLEncoding.EncodeToString(signature)
+	jwt := headerBase64 + "." + payloadBase64 + "." + signatureBase64
+	return &Attestation{
+		PublicKeyID:       s.publicKeyID,
+		Signature:         []byte(jwt),
+		SerializedPayload: payload,
```

Comment:
from attestation.go: // SerializedPayload stores the payload over which the signature was
	// signed. This field is only used for PKIX Attestations.

This is not accurate in the case of JWTs as written now. I think the cleanest way to handle it is to leave SerializedPayload empty for JWTs and update the documentation to reflect that

Reply:
this was a mistake. I updated it to leave the SerializedPayload field empty.

Reply:
done.